### PR TITLE
Ticket #1139: Import additional data for transferred domains

### LINKED ISSUE #1139: Import additional data for transferred domains
### Issue description and context

Analysts need good metadata about domains in-registrar. #923 pulled some data from exports, but we should also include a domain's type, the value for its ["Agency" and "Organization"](https://flatgithub.com/cisagov/dotgov-data/blob/main/?filename=current-full.csv&sha=40ebece364b35581ae2dbf3fe0e3eb5b55c3e2c8), the creation date, and expiration date.

* **Domain types** are defined in _domaintypes.adhoc.dotgov.txt_ (not all values are used). `domaintypeid` is the key, and the domain data is in _domainadditionaldatalink.adhoc.dotgov.txt_. Organization type has **federal type** and **organization type** (requires mapping)

* **Agency** data is at _agency.adhoc.dotgov.txt_. These are the agency names for federal agencies; everyone else is `Non-federal agency`. (I'm unsure where else this key appears.) This will map to pull out **federal_agency**

* **Organization** data is in _domainadditionaldatalink.adhoc.dotgov.txt_,  and `orgid` can be looked up in `20231009.organization.adhoc.dotgov`. 
**has organization name**
**Domain creation and expiration dates** are in `escrow_contacts.daily.dotgov.GOV`, the 8th and 12th column respectively (if useful, I believe the relevant RFC is https://datatracker.ietf.org/doc/html/rfc9022#section-5.1.2.1.1). has **creation date, expiration date **

### Acceptance criteria

- [ ] Transition domain model  is updated to include domain Organization Type, Organization Name, federal agency, federal type, creation date, expiration date 

- [ ] Analysts can easily see what a transferred domain's type is, when it was originally created and when its current expiration date is. 
  - [ ] creation date should be changed on a domain's object in the domain table while expiration is saved in the transition domains table.
  - [ ] domain information should be updated with the correct fields (org type, org name federal agency, federal type) 
- [ ] READMEs are updated accordingly

### Additional context
The .txt files mentioned are in [Drive](https://drive.google.com/drive/folders/1TWxHUZexMqTK6VH9iXgopiX7OknX4vOC?usp=drive_link).
Implementation suggestions not criteria: 
domain type should map to DomainInformation Organization Type
When getting domain type if it is federal, save the Judicial/executive/legistative portion in a new column on transition domains for federal_type
When creating domain information (see 980's implementation on user.py), adjust the DomainInformation creation to include organization type, federal_type (if applicable) and federal_agency (Non-federal agency)
When pulling in federal_agency, find the difference between our current list and the one provided from verisign and add any missing agencies to our AGENCY list defined in Domain Application. 
note on federal agency, if it is a Non-federal, the federal_agency field should just be null

### Links to other issues

🔁 Related to #980 #923

### PR DESCRIPTION
**Pull request by zandercymatics**
## Ticket

Resolves #1139

## Changes
This PR is rather lengthy, so even just a review of one specific file is helpful here. Don't feel as though all of it needs to be passed in one go! If anything is overwhelming or confusing, please consider shooting me a message and I can do a code walkthrough.

### Modified file load_transition_domain.py
- Removed the domain_contacts_filename/contacts_filename as mandatory arguments and replaced it with migration_json_filename which is a JSON file that contains all relevant filenames.
- Added optional arguments for the other filenames (Will be overwritten by the JSON)
- Added an --infer_filenames flag. Intended for use during development only.
- Added a parse_extra function that calls the helper loader class.
- Modified `handle` such that it can accommodate a JSON file for passing settings
- Added some additional error handling for invalid filepaths
### Modified file master_domain_migrations.py
- Added migration_json_filename as a mandatory argument
- Modified file to accommodate the JSON and additional data
- Bug fixes
### Modified transfer_transition_domains_to_domains
- Bug fixes / log cleanup
- update_or_create_domain function that handles adding creation date and expiration date + other data
- update_domain_status function
- create_new_domain_info function that is responsible for adding data and mapping it to existing fields
- update_domain_information function that updates a a DomainInformation object on desired fields
- update_or_create_domain_information which is a container for both of the above
- Bulk creating DomainInformation objects + logging
### Added epp_data_containers.py
- Added dataclasses for AgencyAdhoc, DomainAdditionalData, DomainTypeAdhoc, OrganizationAdhoc, AuthorityAdhoc, DomainEscrow, and an enum EnumFilenames
- These classes are used for organizing and managing data such that it is easier to work with and less error prone
### Added extra_transition_domain_helper.py
- Added classes FileTransitionLog, LoadExtraTransitionDomain, FileDataHolder, ExtraTransitionDomain
### Modified terminal_helper.py
- Added a LogCode enum to better organize log events
- Modified some helper methods
### Added transition_domain_arguments.py
- Added TransitionDomainArguments. Enforces typing/structure while remaining flexible like a dictionary. Designed so that globally tracking variables between files (since they are all interlinked) can be done in one location in a consistent way.
### Modified transition_domain.py
- Added fields organization_type, organization_name, federal_type, federal_agency, epp_creation_date, and epp_expiration_date
### Added Test data files under registrar/tests/data
- Usually about 5 records per file
### Modified test_transition_domain_migrations.py
- Modified tests to accommodate JSON change and deal with additional data. Added tests for DomainInformation / TransitionDomain
### Modifed mixins.py
- Fixed an edge case for the "Manage Domain" button in /admin. Previously, it would assume that a DomainApplication object exists, but since it doesn't for transition records (it, rather, gets created), an authenticated user can now still view the domain. 

## Context for reviewers
Some of this PR still needs some cleanup + test case expansion. Ignore that for now, but the core functionality is testable. 

Download this [file](https://drive.google.com/file/d/1wU_5jxjecyoJ-KNFWMGwMjfI2bfM7N0x/view?usp=sharing) for our data and the given JSON file.

Since we now need to pass in a myriad of filenames, we are tracking those (and the parent directory) inside of a JSON file. The filename will be passed into the script you will call. For ease of testing, we suggest that you use the one that is given inside the JSON file but this can be modified. Here is an example:
```
{
    "directory": "migrationdata/",
    "agency_adhoc_filename": "20231009.agency.adhoc.dotgov.txt",
    "authority_adhoc_filename": "authority.adhoc.dotgov.txt",
    "contacts_filename": "escrow_contacts.daily.dotgov.GOV.txt",
    "domain_adhoc_filename": "20231009.domaintypes.adhoc.dotgov.txt",
    "domain_additional_filename": "20231009.domainadditionaldatalink.adhoc.dotgov.txt",
    "domain_contacts_filename": "escrow_domain_contacts.daily.dotgov.GOV.txt",
    "domain_escrow_filename": "escrow_domains.daily.dotgov.GOV.txt",
    "domain_statuses_filename": "escrow_domain_statuses.daily.dotgov.GOV.txt",
    "organization_adhoc_filename": "20231009.organization.adhoc.dotgov.txt"
}
```

Do note that some of the error handing on the JSON portion is a bit finicky (logs getting buried). That is being worked on. If data is not loading, it is likely due to a wrong filename.

### Testing on getgov-za
All of the data will be preloaded on my sandbox using these scripts. Navigate to /admin and click Domains, TransitionDomains, and DomainInvitations to see this data. We can expect some federal fields to be none if the type is not "federal", but otherwise most of the information should be populated. To verify data integrity manually, you can cross reference the given files noted in the `Testing on your own branch` section. Try messing with this data and accessing it both through invitations and managing a domain.

If you'd like, you can import your own data using the steps documented in a [related PR](https://github.com/cisagov/manage.get.gov/pull/1205).
 
### Testing locally
**IMPORTANT:** Before testing, be sure to do `docker-compose build`. You may also have to apply migrations, given that this file concerns that.

To test this PR accurately, this ticket requires multiple .txt files and a .json file. This is bundled [here](https://drive.google.com/file/d/1wU_5jxjecyoJ-KNFWMGwMjfI2bfM7N0x/view?usp=sharing). For the purposes of this PR, move this file to the `src/migrationdata` directory. **IMPORTANT:** Be careful not to accidentally commit anything. The gitignore file has been temporarily modified to help mitigate this issue. 

After those files have been uploaded, follow these steps (Note that you will see a lot of logs - this is normal):
1. docker-compose exec app ./manage.py load_transition_domain migrationFilepaths.json --debug --limitParse 50
2. docker-compose exec app ./manage.py transfer_transition_domains_to_domains --debug
3. If all is well, navigate to /admin to verify data integrity.
4. docker-compose exec app ./manage.py master_domain_migrations migrationFilepaths.json --runMigrations --resetTable --prompt --debug --limitParse 500
Hit "y" for all prompts minus the last. All emails are currently set to Alysia's.
```
Proceed with sending user invites for all transition domains?
                (Y = proceed, N = skip)
```
^^^^ Hit "N"
5. Verify data integrity on /admin

#### Satisfied acceptance criteria and met development standards

- [x] Met the acceptance criteria, or will meet them in a subsequent PR
- [ ] Created/modified automated tests
- [x] Added at least 2 developers as PR reviewers (only 1 will need to approve)
- [ ] Messaged on Slack or in standup to notify the team that a PR is ready for review
- [ ] Changes to “how we do things” are documented in READMEs and or onboarding guide
- [ ] If any model was updated to modify/add/delete columns, makemigrations was ran and the assoicated migrations file has been commited.

#### Ensured code standards are met (Original Developer)

- [x] All new functions and methods are commented using plain language
- [ ] Did dependency updates in Pipfile also get changed in requirements.txt?
- [x] Interactions with external systems are wrapped in try/except
- [x] Error handling exists for unusual or missing values

#### Validated user-facing changes (if applicable)

- [ ] New pages have been added to .pa11yci file so that they will be tested with our automated accessibility testing
- [ ] Checked keyboard navigability
- [ ] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)
- [ ] Add at least 1 designer as PR reviewer

#### Reviewed, tested, and left feedback about the changes

- [ ] Pulled this branch locally and tested it
- [ ] Reviewed this code and left comments
- [ ] Checked that all code is adequately covered by tests
- [ ] Made it clear which comments need to be addressed before this work is merged
- [ ] If any model was updated to modify/add/delete columns, makemigrations was ran and the assoicated migrations file has been commited.

#### Ensured code standards are met (Code reviewer)

- [ ] All new functions and methods are commented using plain language
- [ ] Interactions with external systems are wrapped in try/except
- [ ] Error handling exists for unusual or missing values
- [ ] (Rarely needed) Did dependency updates in Pipfile also get changed in requirements.txt?

#### Validated user-facing changes as a developer

- [ ] New pages have been added to .pa11yci file so that they will be tested with our automated accessibility testing
- [ ] Checked keyboard navigability
- [ ] Meets all designs and user flows provided by design/product
- [ ] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)

- [ ] Tested with multiple browsers, the suggestion is to use ones that the developer didn't (check off which ones were used)
  - [ ] Chrome
  - [ ] Microsoft Edge
  - [ ] FireFox
  - [ ] Safari

- [ ] (Rarely needed) Tested as both an analyst and applicant user

**Note:** Multiple code reviewers can share the checklists above, a second reviewers should not make a duplicate checklist

#### Verified that the changes match the design intention

- [ ] Checked that the design translated visually
- [ ] Checked behavior
- [ ] Checked different states (empty, one, some, error)
- [ ] Checked for landmarks, page heading structure, and links
- [ ] Tried to break the intended flow

#### Validated user-facing changes as a designer

- [ ] Checked keyboard navigability
- [ ] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)

- [ ] Tested with multiple browsers (check off which ones were used)
  - [ ] Chrome
  - [ ] Microsoft Edge
  - [ ] FireFox
  - [ ] Safari

- [ ] (Rarely needed) Tested as both an analyst and applicant user

### CODE REVIEW

File: src/registrar/management/commands/utility/terminal_helper.py [hunk 1]
```diff
@@ -1,8 +1,25 @@
...
+    - 3 INFO
+    - 4 DEBUG
+    - 5 DEFAULT
+    """
+
+    ERROR = 1
+    WARNING = 2
+    INFO = 3
+    DEBUG = 4
+    DEFAULT = 5
```
Comment (zandercymatics) [inline src/registrar/management/commands/utility/terminal_helper.py:22]: We throw a lot of logs around when running these scripts, especially when parsing additional data. This enumerator allows us to keep tabs on what "kind" of error was thrown. 

Default specifies that no log severity was defined. At that point, it is up to the function to decide how to handle it.

File: .gitignore [hunk 2]
```diff
@@ -171,3 +171,6 @@ node_modules
 
 # Compliance/trestle related
 docs/compliance/.trestle/cache
+
+src/migrationdata/*
```
Comment (zandercymatics) [inline .gitignore:175]: This change is temporary. This is here to avoid accidentally committing anything to the repo while testing

File: src/registrar/management/commands/load_transition_domain.py [hunk 3]
```diff
@@ -312,8 +458,10 @@ def handle(  # noqa: C901
 
         # Start parsing the main file and create TransitionDomain objects
         logger.info("Reading domain-contacts data file %s", domain_contacts_filename)
+        total_lines = TerminalHelper.get_file_line_count(domain_contacts_filename)
         with open(domain_contacts_filename, "r") as domain_contacts_file:
             for row in csv.reader(domain_contacts_file, delimiter=sep):
+                # TerminalHelper.printProgressBar(total_rows_parsed, total_lines)
```
Comment (abroddrick) [inline src/registrar/management/commands/load_transition_domain.py:464]: **question**: was this something just being tried out?

File: src/registrar/management/commands/agency_data_extractor.py [hunk 4]
```diff
@@ -0,0 +1,273 @@
...
+            )
+        if prompt_successful or not prompt:
+            logger.info(
+                f"\n{TerminalColors.OKGREEN}"
+                f"\n======================== MERGED LISTS (transition domain + agency file) ============================"
+                f"\nThese are all the agencies our transition domains table plus all the agencies in the agency file."
+                f"\n\n{len(merged_agencies)} TOTAL\n\n"
+            )
+            self.print_agency_list(
+                merged_transition_agencies, "Merged_Transition_Domain_Agency_List"
```
[the remark this answers is not in the document]
Comment (zandercymatics) [reply src/registrar/management/commands/agency_data_extractor.py]: Deleted

File: src/registrar/views/utility/mixins.py [hunk 5]
```diff
@@ -100,7 +100,14 @@ def can_access_other_user_domains(self, pk):
         if DomainInformation.objects.filter(id=pk).exists():
             requested_domain = DomainInformation.objects.get(id=pk)
 
-        if requested_domain.domain_application.status not in valid_domain_statuses:
+        # If no domain_application object exists and we are
+        # coming from the manage_domain dashboard, this is likely
+        # a transition domain.
+        domain_application = requested_domain.domain_application
```
[the remark this answers is not in the document]
Comment (zandercymatics) [reply src/registrar/views/utility/mixins.py:106]: [Daves PR](https://github.com/cisagov/manage.get.gov/pull/1302) fixes this I believe, its a workaround until then.
This is an area I worked on a while ago...

Basically the manage domain (/admin) button does a check on the status, and these only exist on a domain_application object. While we do create those objects, they don't initially exist from the dashboard so it errors out (because it expects there to be one). 

The code above this checks if the user is authorized, if they clicked the button on the admin dashboard, and if they are navigating to the desired domain. Technically not accurate, but conceptually this is like adding "none" in the list of acceptable statuses

File: src/registrar/management/commands/load_transition_domain.py (hunk 3 above)
Comment (zandercymatics) [reply src/registrar/management/commands/load_transition_domain.py:463]: Yeah I believe so

File: docs/operations/data_migration.md [hunk 6]
```diff
@@ -79,11 +79,36 @@ docker compose run app ./manage.py load_domain_invitations /app/escrow_domain_co
...
 - FILE 3: **escrow_domain_statuses.daily.gov.GOV.txt** -> has the map of domains and their statuses
+- FILE 4: **escrow_domains.daily.dotgov.GOV.txt** -> has basic domain data
+- FILE 5: **domainadditionaldatalink.adhoc.dotgov.txt** -> has the map of domains to other data like authority, organization, & domain type
+- FILE 6: **domaintypes.adhoc.dotgov.txt** -> has domain type data
+- FILE 7: **organization.adhoc.dotgov.txt** -> has organization data
+- FILE 8: **authority.adhoc.dotgov.txt** -> has authority data
+- FILE 9: **agency.adhoc.dotgov.txt** -> has agency data
+
+#### STEP 2: obtain JSON file (for file locations)
+Add a JSON file called "migrationFilepaths.json" with the following contents (update filenames and directory as needed):
```
[the remark this answers is not in the document]
Comment (zandercymatics) [reply docs/operations/data_migration.md:97]: It would need to be relative, yes but it can be anywhere

**Review by abroddrick — APPROVED** [review]
Runs and works

File: src/registrar/models/domain_information.py [hunk 7]
```diff
@@ -54,7 +54,6 @@ class DomainInformation(TimeStampedModel):
         blank=True,
         help_text="Type of Organization",
     )
-
```
Comment (abroddrick) [inline src/registrar/models/domain_information.py:57]: add this line

File: src/registrar/management/commands/utility/extra_transition_domain_helper.py [hunk 8]
```diff
@@ -0,0 +1,1022 @@
...
+        self.data_type = data_type
+
+        # == What the id should be in the holding dict ==#
+        # TODO - rename to id_field_name
+        self.id_field = id_field
+
+        # Object data #
+        self.data: Dict[str, type] = {}
+
+    def try_infer_filename(self, current_file_name, default_file_name):
```
Comment (neilmb) [inline src/registrar/management/commands/utility/extra_transition_domain_helper.py:688]: **Comment (blocking)**: We should never infer filenames. The user who constructs migration_filepaths.json must give the right ones. This is too important to guess. Please pull all of this out. No regexes, no dates, just give the right filename in the JSON input.

File: src/registrar/management/commands/utility/extra_transition_domain_helper.py [hunk 9]
```diff
@@ -0,0 +1,1022 @@
...
+            )
+        return file_data
+
+    def parse_all_files(self, infer_filenames=True):
+        """Clears all preexisting data then parses each related CSV file.
+
+        infer_filenames: bool -> Determines if we should try to
+        infer the filename if a default is passed in
+        """
+        self.clear_file_data()
```
Comment (neilmb) [inline src/registrar/management/commands/utility/extra_transition_domain_helper.py:853]: **question (non-blocking)**: This is only ever called from `__init__`. Why do we need to clear data here?

File: src/registrar/management/commands/utility/terminal_helper.py [hunk 10]
```diff
@@ -98,5 +185,75 @@ def prompt_for_execution(
...
+            # Write to file
+            logger.info(
+                f"{TerminalColors.MAGENTA}Writing to file "
+                f" {filepath}..."
+                f"{TerminalColors.ENDC}"
+            )
+            with open(f"{filepath}", "w+") as f:
+                f.write(file_contents)
+
+    @staticmethod
```
Comment (neilmb) [inline src/registrar/management/commands/utility/terminal_helper.py:217]: **comment (blocking)**: This is never called, please remove dead code.

**Review by neilmb — APPROVED** [review]
Okay, sure, but let's at least get rid of the code that you aren't using. The test cases are nice to see. The docs and the command line options are a mess. The data model between extra domains and all the dataclasses is very hard to follow and I'm hoping we'll be able to figure out how to add AO and address information by following this.

File: src/registrar/management/commands/utility/extra_transition_domain_helper.py (hunk 8 above)
Comment (zandercymatics) [reply src/registrar/management/commands/utility/extra_transition_domain_helper.py:688]: The inference stuff is a tool for debugging / local development but not used on the script normally (so I can hot swap old data for the latest). This only occurs when the flag --infer_filenames is passed in which is defaulted to false. With that optional bool off, it will throw an error if it can't find the filename normally.
 
 Should this still be removed? In practice this action will never occur unless specified. I can also just hard code it such that the flag is always false and not give the end user the option to even pass it in, which is fine since its just for development. I personally would opt for that approach.

File: src/registrar/management/commands/utility/extra_transition_domain_helper.py (hunk 9 above)
Comment (zandercymatics) [reply src/registrar/management/commands/utility/extra_transition_domain_helper.py:853]: Good call. The original intent was to allow the script to attempt to conditionally re-parse under certain conditions (like renaming a file on the fly) without having to reload the entire script but that didn't get implemented. Code remnant

File: src/registrar/management/commands/utility/extra_transition_domain_helper.py (hunk 8 above)
Comment (neilmb) [reply src/registrar/management/commands/utility/extra_transition_domain_helper.py:662]: **comment**: I'm fine with only doing this in DEBUG environments, that feels very safe.

**Review by neilmb — APPROVED** [review]
This is improved @zandercymatics, thanks for your hard work on it. Merge at will and we'll get it released to staging and stable.
